ibm/resource_ibm_cis_dns_record.go

```diff
@@ -0,0 +1,540 @@
+	cisId := d.Get("cis_id").(string)
+	zoneId := d.Get("domain_id").(string)
+
+	newRecord := v1.DnsBody{
+		DnsType: d.Get("type").(string),
+		Name:    d.Get("name").(string),
+	}
+
+	content, contentOk := d.GetOk("content")
+	if contentOk {
+		newRecord.Content = content.(string)
+	}
+
+	data, dataOk := d.GetOk("data")
+	log.Printf("[DEBUG] Data found in config: %#v", data)
+
+	newDataMap := make(map[string]interface{})
+
+	if dataOk {
+		for id, content := range data.(map[string]interface{}) {
+			newData, err := transformToIBMCISDNSDNSData(newRecord.DnsType, id, content)
+			if err != nil {
+				return err
+			} else if newData == nil {
+				continue
+			}
+			newDataMap[id] = newData
+		}
+
+		newRecord.Data = newDataMap
+	}
+
+	if contentOk == dataOk {
+		return fmt.Errorf(
+			"either 'content' (present: %t) or 'data' (present: %t) must be provided",
+			contentOk, dataOk)
+	}
+
+	if priority, ok := d.GetOk("priority"); ok {
+		newRecord.Priority = priority.(int)
+	}
+
+	// Validate content based on type
+	if err := validateRecordName(newRecord.DnsType, newRecord.Content); err != nil {
+		return fmt.Errorf("Error validating record name %q: %s", newRecord.Name, err)
+	}
+
+	log.Printf("[DEBUG] IBMCISDNS Record create configuration: %#v", newRecord)
+
+	var recordPtr *v1.DnsRecord
+	recordPtr, err = cisClient.Dns().CreateDns(cisId, zoneId, newRecord)
+	if err != nil {
+		return fmt.Errorf("Failed to create record: %s", err)
+	}
+
+	// In the Event that the API returns an empty DNS Record, we verify that the
+	// ID returned is not the default ""
+
+	r := *recordPtr
+
+	if r.Id == "" {
+		return fmt.Errorf("Failed to find record in Create response; Record was empty")
+	}
+
+	d.SetId(r.Id)
+
+	log.Printf("[INFO] IBMCISDNS Record Id: %s", d.Id())
+
+	return resourceIBMCISDnsRecordRead(d, meta)
+}
+
+func resourceIBMCISDnsRecordRead(d *schema.ResourceData, meta interface{}) error {
+	cisClient, err := meta.(ClientSession).CisAPI()
+	zoneId := d.Get("domain_id").(string)
+	cisId := d.Get("cis_id").(string)
+
+	var recordPtr *v1.DnsRecord
+	recordPtr, err = cisClient.Dns().GetDns(cisId, zoneId, d.Id())
+	if err != nil {
+		if strings.Contains(err.Error(), "Invalid dns record identifier") ||
+			strings.Contains(err.Error(), "HTTP status 404") {
+			log.Printf("[WARN] Removing record from state because it's not found in API")
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	record := *recordPtr
+	//d.SetId(record.Id)
+	d.Set("name", record.Name)
+	d.Set("type", record.DnsType)
+	d.Set("content", record.Content)
+	d.Set("ttl", record.Ttl)
+	d.Set("priority", record.Priority)
+	d.Set("proxied", record.Proxied)
+	d.Set("created_on", record.CreatedOn.Format(time.RFC3339Nano))
+	d.Set("data", expandStringMap(record.Data))
+	d.Set("modified_on", record.ModifiedOn.Format(time.RFC3339Nano))
+	d.Set("proxiable", record.Proxiable)
+
+	return nil
+}
+
+func resourceIBMCISDnsRecordUpdate(d *schema.ResourceData, meta interface{}) error {
```

Comment:
Can we remove this update if nothing is being updated?

Reply:
Place holder for required future update function.